lib/Sema/SemaBounds.cpp

```diff
@@ -578,6 +578,243 @@ namespace {
 
       return true;
     }
+
+    // This includes both ImplicitCastExprs and CStyleCastExprs
+    bool VisitCastExpr(CastExpr *E) {
+      CheckDisallowedFunctionPtrCasts(E);
+
+      return true;
+    }
+
+  private:
+    // Here we're examining places where a programmer has cast to a
+    // checked function pointer type, in order to make sure this cast is
+    // safe and valid.
+    // Let's term the outer value (after the cast), E, of type ToType.
+    // In the values we're examining, ToType is a ptr<> to a function type.
+    //
+    // To produce E, the programmer is performing a sequence of casts,
+    // both implicit and explicit, and perhaps this sequence includes using
+    // addr-of (&) or deref(*).
+    //
+    // We search this chain, starting at E.
+    //
+    // 1. If we come across something of ptr<> function type, then one of two things
```

Comment:
It would be good to update the comment to match the order of the code below.  In particular, we handle null pointers first.    That way the code doesn't have to cite an out-of-order case.